src/Middleware/RateLimiting/src/RateLimitingMiddleware.cs

```diff
@@ -44,30 +60,124 @@ public RateLimitingMiddleware(RequestDelegate next, ILogger<RateLimitingMiddlewa
+    }
+
+    private LeaseContext CombinedAcquire(HttpContext context)
+    {
+        RateLimitLease? globalLease = null;
+        RateLimitLease? endpointLease = null;
+
+        try
+        {
+            if (_globalLimiter is not null)
+            {
+                globalLease = _globalLimiter.Acquire(context);
+                if (!globalLease.IsAcquired)
+                {
+                    return new LeaseContext() { GlobalRejected = true, Lease = globalLease };
+                }
+            }
+            endpointLease = _endpointLimiter.Acquire(context);
+            if (!endpointLease.IsAcquired)
+            {
+                globalLease?.Dispose();
+                return new LeaseContext() { GlobalRejected = false, Lease = endpointLease };
+            }
+        }
+        catch (Exception)
+        {
+            endpointLease?.Dispose();
+            globalLease?.Dispose();
+            throw;
+        }
+        return globalLease is null ? new LeaseContext() { Lease = endpointLease } : new LeaseContext() { Lease = new DefaultCombinedLease(globalLease, endpointLease) };
+    }
+
+    private async ValueTask<LeaseContext> CombinedWaitAsync(HttpContext context, CancellationToken cancellationToken)
+    {
+        RateLimitLease? globalLease = null;
+        RateLimitLease? endpointLease = null;
+
+        try
+        {
+            if (_globalLimiter is not null)
+            {
+                globalLease = await _globalLimiter.WaitAsync(context, cancellationToken: cancellationToken);
+                if (!globalLease.IsAcquired)
+                {
+                    return new LeaseContext() { GlobalRejected = true, Lease = globalLease };
+                }
+            }
+            endpointLease = await _endpointLimiter.WaitAsync(context, cancellationToken: cancellationToken);
+            if (!endpointLease.IsAcquired)
+            {
+                globalLease?.Dispose();
+                return new LeaseContext() { GlobalRejected = false, Lease = endpointLease };
+            }
+        }
+        catch (Exception)
+        {
+            endpointLease?.Dispose();
+            globalLease?.Dispose();
+            throw;
+        }
+
+        return globalLease is null ? new LeaseContext() { Lease = endpointLease } : new LeaseContext() { Lease = new DefaultCombinedLease(globalLease, endpointLease) };
+    }
+
+    // Create the endpoint-specific PartitionedRateLimiter
+    private PartitionedRateLimiter<HttpContext> CreateEndpointLimiter()
+    {
+        // If we have a policy for this endpoint, use its partitioner. Else use a NoLimiter.
+        return PartitionedRateLimiter.Create<HttpContext, DefaultKeyType>(context =>
+        {
+            var name = context.GetEndpoint()?.Metadata.GetMetadata<IRateLimiterMetadata>()?.PolicyName;
+            if (name is not null && _policyMap.TryGetValue(name, out var policy))
+            {
+                return policy.GetPartition(context);
+            }
```

Comment:
Can we log something if `name is not null` but it's not in the `_policyMap`? That should at least be an error, right?

Reply:
That would mean a user called `RequireRateLimiting` on an endpoint w/ some policy name, but did not actually add any policy with that name. That feels more like a warning to me - I'll add one

Reply:
Actually I'll make it an exception